sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -84,24 +86,125 @@ static RT::PiProgram createSpirvProgram(const RT::PiContext Context,
+
+  KernelSetId KSId = getKernelSetId(M, KernelName);
+  return getDeviceImage(M, KSId, Context);
+}
+
+static bool is_device_binary_type_supported(const context &C,
+                                            RT::PiDeviceBinaryType Format) {
+  // All formats except PI_DEVICE_BINARY_TYPE_SPIRV are supported.
+  if (Format != PI_DEVICE_BINARY_TYPE_SPIRV)
+    return true;
+
+  // OpenCL 2.1 and greater require clCreateProgramWithIL
+  if (pi::useBackend(pi::SYCL_BE_PI_OPENCL) &&
+      C.get_platform().get_info<info::platform::version>() >= "2.1")
+    return true;
+
+  // Otherwise we need cl_khr_il_program extension to be present
+  // and we can call clCreateProgramWithILKHR using the extension
+  for (const auto &D : C.get_devices()) {
+    auto Extensions = D.get_info<info::device::extensions>();
+    if (std::find(Extensions.begin(), Extensions.end(),
+                  string_class("cl_khr_il_program")) != Extensions.end())
+      return true;
+  }
+
+  // This device binary type is not supported.
+  return false;
+}
+
+RT::PiProgram ProgramManager::createPIProgram(const DeviceImage &Img,
+                                              const context &Context) {
+  if (DbgProgMgr > 0)
+    std::cerr << ">>> ProgramManager::createPIProgram(" << &Img << ")\n";
+
+  // perform minimal sanity checks on the device image and the descriptor
+  if (Img.BinaryEnd < Img.BinaryStart) {
+    throw runtime_error("Malformed device program image descriptor");
+  }
+  if (Img.BinaryEnd == Img.BinaryStart) {
+    throw runtime_error("Invalid device program image: size is zero");
+  }
+  size_t ImgSize = static_cast<size_t>(Img.BinaryEnd - Img.BinaryStart);
+
+  // TODO if the binary image is a part of the fat binary, the clang
+  //   driver should have set proper format option to the
+  //   clang-offload-wrapper. The fix depends on AOT compilation
+  //   implementation, so will be implemented together with it.
+  //   Img->Format can't be updated as it is inside of the in-memory
+  //   OS module binary.
+  auto Format = getFormat(Img);
+  // auto Format = Img->Format;
+  // assert(Format != PI_DEVICE_BINARY_TYPE_NONE && "Image format not set");
+
+  // Dump program image if requested
+  if (std::getenv("SYCL_DUMP_IMAGES") && SpvFile.empty()) {
+    std::string Fname("sycl_");
+    Fname += Img.DeviceTargetSpec;
+    std::string Ext;
+
+    if (Format == PI_DEVICE_BINARY_TYPE_SPIRV)
+      Ext = ".spv";
+    else if (Format == PI_DEVICE_BINARY_TYPE_LLVMIR_BITCODE)
+      Ext = ".bc";
+    else
+      Ext = ".bin";
+    Fname += Ext;
+
+    std::ofstream F(Fname, std::ios::binary);
+
+    if (!F.is_open()) {
+      throw runtime_error(std::string("Can not write ") + Fname);
+    }
+    F.write(reinterpret_cast<const char *>(Img.BinaryStart), ImgSize);
+    F.close();
+  }
+
+  if (!is_device_binary_type_supported(Context, Format))
+    throw feature_not_supported(
+        "Online compilation is not supported in this context");
```

Comment:
This error message makes assumption that the only reason why we could reject binary is a lack of support for SPIRV compilation. Please, consider throwing exception in the is_device_binary_type_supported where the error context is known.

Reply:
I'd prefer to address this separately.

Reply:
Ok